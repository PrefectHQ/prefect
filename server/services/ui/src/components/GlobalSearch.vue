<script>
import globalNameSearchQuery from '@/graphql/GlobalSearch/search-by-name.gql'
import globalIDSearchQuery from '@/graphql/GlobalSearch/search-by-id.gql'
import GlobalSearchIcon from '@/components/GlobalSearchIcon'
import GlobalSearchResult from '@/components/GlobalSearchResult'

export default {
  components: { GlobalSearchIcon, GlobalSearchResult },
  data() {
    return {
      input: null,
      model: null,
      search: null,
      results: [],
      isLoading: false
    }
  },
  computed: {
    id() {
      if (!this.input) return ''

      // Call .trim() to get rid of whitespace on the ends of the
      // string before making the query
      return `${this.input.trim()}`
    },
    isSearchForID() {
      const UUIDRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/

      // Call .trim() to get rid of whitespace on the ends of the
      // string before making the query
      return UUIDRegex.test(this.input.trim())
    },
    name() {
      if (!this.input) return ''
      // Call .trim() to get rid of whitespace on the ends of the
      // string before making the query
      return `%${this.input.trim()}%`
    },
    entries() {
      return this.results
    }
  },
  watch: {
    search(val) {
      this.handleSearch(val)
    },
    model(val) {
      this.handleResultSelected(val)
    }
  },
  mounted() {
    // Adds the event listener for the / search shortcut
    window.addEventListener('keyup', this.handleKeyboardShortcut)
  },
  beforeDestroy() {
    // Removes the / search shortcut event listener when
    // the component is destroyed
    window.removeEventListener('keyup', this.handleKeyboardShortcut)
  },
  methods: {
    handleKeyboardShortcut(e) {
      if (
        e.key &&
        e.key === '/' &&
        e.srcElement.tagName !== 'INPUT' &&
        e.srcElement.tagName !== 'TEXTAREA'
      ) {
        // We use the .onClick() method here rather than .focus()
        // because .focus() doesn't bring up the context menu
        // .onClick() also calls .focus()
        this.$refs['global-search'].onClick()
      }
    },
    handleSearch(input) {
      // We set the input prop to null if the input
      // string is empty, to bring back up the searchbox hint
      this.input = input == '' ? null : input

      // If there's no input or apollo has been told to
      // pause all queries elsewhere (the skipAll prop),
      // we pause the individual queries and
      // remove all results. This will show the typing hint
      // again if the user has focused the global search focus
      if (!input || this.$apollo.skipAll) {
        this.$apollo.queries.nameQuery.skip = true
        this.$apollo.queries.idQuery.skip = true
        this.results = []
        return
      }

      // Once we've confirmed that we have input and
      // we aren't pausing all queries, we check that
      // the input matches a UUID regex which is stored as
      // a computed prop called isSearchForID
      // If that's the case, we start an id query and
      // pause the name query.
      if (this.isSearchForID) {
        this.startQuery('id')
        this.$apollo.queries.nameQuery.skip = true
      }
      // Otherwise, we start a name query and pause
      // the id query.
      else {
        this.startQuery('name')
        this.$apollo.queries.idQuery.skip = true
      }
    },
    handleResultSelected(searchResultName) {
      // If this is called with no argument
      // we return immediately
      if (!searchResultName) return

      // The search result that we use is found in the
      // entries on name, since the VAutocomplete
      // component will always pass the name prop to this
      // method regardless of the query used (id or name)
      let searchResult = this.entries.find(
        result => result.name == searchResultName
      )

      // Pause all queries
      this.$apollo.skipAll = true

      const routeToNavigateTo = this.routeName(searchResult.__typename)

      // Navigate to the URL based on the search result
      this.$router.push({
        name: routeToNavigateTo,
        params: { id: searchResult.id }
      })
    },
    routeName(name) {
      switch (name) {
        case 'flow':
          return 'flow'
        case 'task':
          return 'task'
        case 'flow_run':
          return 'flow-run'
        case 'task_run':
          return 'task-run'
        default:
          throw new Error('Unable to resolve route, GlobalSearchResult')
      }
    },
    startQuery(ref) {
      this.isLoading = true
      // We make sure to unpause this query before
      // we try to refetch the data
      this.$apollo.queries[`${ref}Query`].skip = false
      this.$apollo.queries[`${ref}Query`].refetch()
    },
    processResult(data, loading) {
      this.isLoading = loading
      // Returning if GraphQL is loading the query still
      // leads to a better UX since the search results
      // don't flicker between loading states
      // i.e. we maintain is loading until GraphQL
      // has determined loading is finished NOT
      // when we've received some data
      if (this.isLoading) return

      this.results = data
        ? Object.entries(data)
            .map(e => e[1].map(e1 => (e1 ? { ...e1 } : [])))
            .flat()
        : []
    },
    searchFilter(item, queryText) {
      // This is the filter we use to determine what the VAutocomplete
      // method is showing. We transform all queries to lowercase
      // for comparison for a better UX
      return (
        item['id'].toLowerCase().includes(queryText.toLowerCase()) ||
        item['name'].toLowerCase().includes(queryText.toLowerCase())
      )
    }
  },
  apollo: {
    idQuery: {
      query: globalIDSearchQuery,
      manual: true,
      debounce: 300,
      fetchPolicy: 'no-cache',
      variables() {
        return { id: this.id }
      },
      result({ data, loading }) {
        this.processResult(data, loading)
      },
      skip: true
    },
    nameQuery: {
      query: globalNameSearchQuery,
      manual: true,
      debounce: 300,
      fetchPolicy: 'no-cache',
      variables() {
        return { name: this.name }
      },
      result({ data, loading }) {
        this.processResult(data, loading)
      },
      skip: true
    }
  }
}
</script>

<template>
  <v-autocomplete
    ref="global-search"
    v-model="model"
    single-line
    dense
    dark
    clearable
    :items="entries"
    :loading="isLoading ? 'green accent-3' : false"
    :search-input.sync="search"
    :filter="searchFilter"
    label="Search"
    placeholder="Search"
    prepend-inner-icon="search"
    item-text="name"
    clear-icon="close"
    open-on-clear
    class="global-search"
    @clear="results = []"
    @focus="
      results = []
      handleSearch(input)
    "
  >
    <template v-if="input == null" v-slot:no-data>
      <v-list-item>
        <v-list-item-title>
          Type to search for a <strong>Flow</strong>, <strong>Task</strong>, or
          <strong>Run</strong>
        </v-list-item-title>
      </v-list-item>
    </template>
    <template v-else-if="isLoading" v-slot:no-data>
      <v-list-item>
        <v-list-item-title>
          Searching...
        </v-list-item-title>
      </v-list-item>
    </template>
    <template v-else v-slot:no-data>
      <v-list-item>
        <v-list-item-title>
          No results matched your search.
        </v-list-item-title>
      </v-list-item>
    </template>
    <template v-slot:item="data">
      <GlobalSearchIcon v-if="data" :type="data.item.__typename" />
      <GlobalSearchResult
        v-if="data"
        :search-result="data.item"
        :parent="data.parent"
      />
    </template>
  </v-autocomplete>
</template>

<style lang="scss" scoped>
$width: 25rem;

.global-search {
  font-size: 1rem;
  height: 2rem;
  max-width: $width;
}

.v-select-list {
  max-width: $width;
}
</style>
