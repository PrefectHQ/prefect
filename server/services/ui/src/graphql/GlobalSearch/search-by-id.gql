query GlobalSearchByID($id: uuid!) {
  flow(where: { id: { _eq: $id } }) {
    id
    name
  }
  flow_run(where: { id: { _eq: $id } }) {
    id
    name
    flow {
      id
      name
    }
  }
  task(where: { id: { _eq: $id } }) {
    id
    name
    flow {
      id
      name
    }
  }
  task_run(where: { id: { _eq: $id } }) {
    id
    task {
      id
      name
      flow {
        id
        name
      }
    }
  }
}
