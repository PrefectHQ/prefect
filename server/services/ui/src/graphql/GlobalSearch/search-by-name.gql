query GlobalSearchByName($name: String) {
  flow(where: { name: { _ilike: $name }, archived: { _eq: false } }, limit: 5) {
    id
    name
  }
  flow_run(
    where: { name: { _ilike: $name }, flow: { archived: { _eq: false } } }
    limit: 5
  ) {
    id
    name
    flow {
      id
      name
    }
  }
  task(
    where: { name: { _ilike: $name }, flow: { archived: { _eq: false } } }
    limit: 5
  ) {
    id
    name
    flow {
      id
      name
    }
  }
}
