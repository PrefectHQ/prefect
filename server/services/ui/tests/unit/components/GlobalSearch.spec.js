// Libraries
import Vuetify from 'vuetify'
import Vue from 'vue'

// Components
import GlobalSearch from '@/components/GlobalSearch'

// Utilities
import { createLocalVue, shallowMount } from '@vue/test-utils'
// import { routes } from '@/router'

// Test Data
// import testData from './GlobalSearchTestData'

// Use jest to mock the graphql data
jest.mock('@/graphql/GlobalSearch/search-by-id.gql', () => {
  return 'a graphql string'
})
jest.mock('@/graphql/GlobalSearch/search-by-name.gql', () => {
  return 'a graphql string'
})

const localVue = createLocalVue()
Vue.use(Vuetify)

describe('GlobalSearch', () => {
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  let wrapper = shallowMount(GlobalSearch, {
    localVue,
    vuetify
  })

  it('renders a vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('focuses the input box when the / hotkey is pressed', () => {
    wrapper.trigger('click')
    wrapper.trigger('keyup', {
      key: '/'
    })

    // Need to add these tests but Jest throws errors
    // when trying to do a full mount as required by Vuetify
    // and I'm not sure how to test vuetify UI components
    // without doing the full vuetify mount
    expect(true).toBe(true)
  })
})
