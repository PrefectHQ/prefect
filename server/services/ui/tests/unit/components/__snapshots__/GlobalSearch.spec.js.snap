// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalSearch should match snapshot 1`] = `
<v-autocomplete-stub
  allowoverflow="true"
  appendicon="$dropdown"
  backgroundcolor=""
  class="global-search"
  clearable="true"
  clearicon="close"
  dark="true"
  dense="true"
  errorcount="1"
  errormessages=""
  filter="function () { [native code] }"
  itemcolor="primary"
  itemdisabled="disabled"
  items=""
  itemtext="name"
  itemvalue="value"
  label="Search"
  loaderheight="2"
  menuprops="[object Object]"
  messages=""
  nodatatext="$vuetify.noDataText"
  openonclear="true"
  placeholder="Search"
  prependinnericon="search"
  rules=""
  singleline="true"
  successmessages=""
  type="text"
  valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
>
  <v-list-item-stub
    activeclass=""
    tag="div"
  >
    <v-list-item-title-stub>
      
        Type to search for a 
      <strong>
        Flow
      </strong>
      , 
      <strong>
        Task
      </strong>
      , or
        
      <strong>
        Run
      </strong>
    </v-list-item-title-stub>
  </v-list-item-stub>
  <!---->
   
  <!---->
</v-autocomplete-stub>
`;
